File: Tests/ConnectorControlCoreTests/RemoteAuthTests.swift
import XCTest
@testable import ConnectorControlCore

final class RemoteAuthTests: XCTestCase {
    private let url = "https://x.dev/mcp"

    private func args(_ v: JSONValue) -> [String]? {
        guard case .object(let o) = v, case .array(let a)? = o["args"] else { return nil }
        return a.compactMap { if case .string(let s) = $0 { return s }; return nil }
    }

    private func env(_ v: JSONValue) -> [String: String] {
        guard case .object(let o) = v, case .object(let e)? = o["env"] else { return [:] }
        var out: [String: String] = [:]
        for (k, val) in e { if case .string(let s) = val { out[k] = s } }
        return out
    }

    // MARK: encode/decode round-trip, one per auth mode

    func testAutomaticRoundTrips() {
        let rc = RemoteConfig(url: url, auth: .automatic)
        let encoded = RemotePattern.encode(rc)
        XCTAssertEqual(args(encoded), ["-y", "mcp-remote", url])
        XCTAssertEqual(RemotePattern.decode(encoded), rc)
    }

    func testBearerRoundTrips() {
        let rc = RemoteConfig(url: url, auth: .bearer(token: "secret-tok"))
        let encoded = RemotePattern.encode(rc)
        XCTAssertEqual(args(encoded),
                       ["-y", "mcp-remote", url, "--header", "Authorization:${AUTH_HEADER}"])
        XCTAssertEqual(env(encoded), ["AUTH_HEADER": "Bearer secret-tok"])
        XCTAssertEqual(RemotePattern.decode(encoded), rc)
    }

    func testHeaderRoundTrips() {
        let rc = RemoteConfig(url: url, auth: .header(name: "X-API-Key", value: "k123"))
        let encoded = RemotePattern.encode(rc)
        XCTAssertEqual(args(encoded),
                       ["-y", "mcp-remote", url, "--header", "X-API-Key:${AUTH_HEADER}"])
        XCTAssertEqual(env(encoded), ["AUTH_HEADER": "k123"])
        XCTAssertEqual(RemotePattern.decode(encoded), rc)
    }

    func testOAuthClientRoundTrips() {
        let rc = RemoteConfig(url: url,
                               auth: .oauthClient(clientID: "cid", clientSecret: "csecret",
                                                   scopes: "read write"))
        let encoded = RemotePattern.encode(rc)
        XCTAssertEqual(args(encoded), [
            "-y", "mcp-remote", url,
            "--static-oauth-client-info", "{\"client_id\":\"cid\",\"client_secret\":\"csecret\"}",
            "--static-oauth-client-metadata", "{\"scope\":\"read write\"}",
        ])
        XCTAssertEqual(RemotePattern.decode(encoded), rc)
    }

    // MARK: specific decode scenarios

    func testDecodeBearerFromExplicitConfig() {
        let config = JSONValue.object([
            "command": .string("npx"),
            "args": .array([.string("-y"), .string("mcp-remote"), .string(url),
                             .string("--header"), .string("Authorization:${AUTH_HEADER}")]),
            "env": .object(["AUTH_HEADER": .string("Bearer abc")]),
        ])
        let decoded = RemotePattern.decode(config)
        XCTAssertEqual(decoded?.auth, .bearer(token: "abc"))
        XCTAssertEqual(decoded?.passthroughEnv, [:])
        // Re-encoding reproduces the same args/env (modulo key order, which
        // JSONValue.object equality already ignores).
        XCTAssertEqual(RemotePattern.encode(decoded!), config)
    }

    func testDecodeCustomHeaderFromExplicitConfig() {
        let config = JSONValue.object([
            "command": .string("npx"),
            "args": .array([.string("-y"), .string("mcp-remote"), .string(url),
                             .string("--header"), .string("X-API-Key:${AUTH_HEADER}")]),
            "env": .object(["AUTH_HEADER": .string("topsecret")]),
        ])
        let decoded = RemotePattern.decode(config)
        XCTAssertEqual(decoded?.auth, .header(name: "X-API-Key", value: "topsecret"))
        XCTAssertEqual(RemotePattern.encode(decoded!), config)
    }

    func testOAuthClientWithoutScopes() {
        let rc = RemoteConfig(url: url,
                               auth: .oauthClient(clientID: "cid", clientSecret: "", scopes: ""))
        let encoded = RemotePattern.encode(rc)
        XCTAssertEqual(args(encoded), [
            "-y", "mcp-remote", url,
            "--static-oauth-client-info", "{\"client_id\":\"cid\",\"client_secret\":\"\"}",
        ])
        XCTAssertEqual(RemotePattern.decode(encoded), rc)
    }

    // MARK: preservation

    func testBearerPlusExtraHeaderPreservesBoth() {
        // A bearer auth (our sentinel pattern) plus a hand-added literal header.
        // The bearer must decode as auth; the extra header must survive verbatim
        // as an extra arg — not clobber the bearer or steal its env var.
        let config = JSONValue.object([
            "command": .string("npx"),
            "args": .array([.string("-y"), .string("mcp-remote"),
                            .string("https://x.dev/mcp"),
                            .string("--header"), .string("Authorization:${AUTH_HEADER}"),
                            .string("--header"), .string("X-Tenant:acme")]),
            "env": .object(["AUTH_HEADER": .string("Bearer secret-tok")]),
        ])
        let decoded = RemotePattern.decode(config)
        XCTAssertEqual(decoded?.auth, .bearer(token: "secret-tok"))
        XCTAssertEqual(decoded?.extraArgs, ["--header", "X-Tenant:acme"])
        XCTAssertTrue(decoded?.passthroughEnv.isEmpty ?? false,
                      "the bearer's AUTH_HEADER is consumed; nothing else leaks")
        // Re-encode keeps both headers.
        let re = RemotePattern.encode(decoded!)
        let a = args(re) ?? []
        XCTAssertTrue(a.contains("Authorization:${AUTH_HEADER}"))
        XCTAssertTrue(a.contains("X-Tenant:acme"))
        XCTAssertEqual(env(re)["AUTH_HEADER"], "Bearer secret-tok")
    }

    func testUnrecognizedHeaderEnvVarIsPreservedNotFabricated() {
        // A header using a NON-sentinel env var must be left as an extra arg and
        // its env var kept in passthrough — never resolved into the auth slot.
        let config = JSONValue.object([
            "command": .string("npx"),
            "args": .array([.string("-y"), .string("mcp-remote"),
                            .string("https://x.dev/mcp"),
                            .string("--header"), .string("X-Key:${MY_KEY}")]),
            "env": .object(["MY_KEY": .string("abc")]),
        ])
        let decoded = RemotePattern.decode(config)
        XCTAssertEqual(decoded?.auth, .automatic)
        XCTAssertEqual(decoded?.extraArgs, ["--header", "X-Key:${MY_KEY}"])
        XCTAssertEqual(decoded?.passthroughEnv["MY_KEY"], "abc")
        XCTAssertEqual(RemotePattern.encode(decoded!), config)
    }

    func testExtraArgsPreserved() {
        let rc = RemoteConfig(url: url, auth: .automatic, extraArgs: ["--transport", "http-only"])
        let encoded = RemotePattern.encode(rc)
        XCTAssertEqual(args(encoded), ["-y", "mcp-remote", url, "--transport", "http-only"])
        XCTAssertEqual(RemotePattern.decode(encoded), rc)
    }

    func testPassthroughEnvSurvivesDecodeThenEncode() {
        let config = JSONValue.object([
            "command": .string("npx"),
            "args": .array([.string("-y"), .string("mcp-remote"), .string(url)]),
            "env": .object(["UNRELATED": .string("keep-me")]),
        ])
        let decoded = RemotePattern.decode(config)
        XCTAssertEqual(decoded?.passthroughEnv, ["UNRELATED": "keep-me"])
        XCTAssertEqual(decoded?.auth, .automatic)
        XCTAssertEqual(RemotePattern.encode(decoded!), config)
    }

    // MARK: versioned marker

    func testVersionedMarkerDecodes() {
        let config = JSONValue.object([
            "command": .string("npx"),
            "args": .array([.string("-y"), .string("mcp-remote@latest"), .string(url)]),
        ])
        let decoded = RemotePattern.decode(config)
        XCTAssertEqual(decoded?.url, url)
        XCTAssertEqual(decoded?.auth, .automatic)
    }

    // MARK: rejection

    func testDecodeReturnsNilForNonNpxCommand() {
        let config = JSONValue.object([
            "command": .string("node"),
            "args": .array([.string("-y"), .string("mcp-remote"), .string(url)]),
        ])
        XCTAssertNil(RemotePattern.decode(config))
    }

    func testDecodeReturnsNilForNonMarker() {
        let config = JSONValue.object([
            "command": .string("npx"),
            "args": .array([.string("-y"), .string("some-other-package"), .string(url)]),
        ])
        XCTAssertNil(RemotePattern.decode(config))
    }

    func testDecodeReturnsNilForMissingURL() {
        let config = JSONValue.object([
            "command": .string("npx"),
            "args": .array([.string("-y"), .string("mcp-remote")]),
        ])
        XCTAssertNil(RemotePattern.decode(config))
    }
}

File: windows/tests/ConnectorControl.Core.Tests/RemoteAuthTests.cs
namespace ConnectorControl.Core.Tests;

public class RemoteAuthTests
{
    private const string Url = "https://x.dev/mcp";

    private static string[]? Args(JsonValue v) =>
        v["args"] is { Kind: JsonKind.Array } a ? a.ArrayItems.Where(i => i.Kind == JsonKind.String).Select(i => i.StringValue).ToArray() : null;

    private static Dictionary<string, string> Env(JsonValue v)
    {
        var result = new Dictionary<string, string>(StringComparer.Ordinal);
        if (v["env"] is { Kind: JsonKind.Object } e)
        {
            foreach (var (k, val) in e.ObjectProperties)
            {
                if (val.Kind == JsonKind.String) { result[k] = val.StringValue; }
            }
        }
        return result;
    }

    private static Dictionary<string, string> Pairs(params (string, string)[] pairs) =>
        pairs.ToDictionary(p => p.Item1, p => p.Item2, StringComparer.Ordinal);

    private static JsonValue NpxConfig(string[] args, Dictionary<string, string>? env = null)
    {
        var props = new List<(string, JsonValue)>
        {
            ("command", JsonValue.String("npx")),
            ("args", JsonValue.Array(args.Select(JsonValue.String))),
        };
        if (env is not null)
        {
            props.Add(("env", JsonValue.Object(env.Select(kv => new KeyValuePair<string, JsonValue>(kv.Key, JsonValue.String(kv.Value))))));
        }
        return JsonValue.Object(props.ToArray());
    }

    // encode/decode round-trip, one per auth mode

    [Fact]
    public void AutomaticRoundTrips()
    {
        var rc = new RemoteConfig(Url, RemoteAuth.Auto, RemoteLaunchStyle.Npx);
        var encoded = RemotePattern.Encode(rc);
        Assert.Equal(["-y", "mcp-remote", Url], Args(encoded)!);
        Assert.Equal(rc, RemotePattern.Decode(encoded));
    }

    [Fact]
    public void BearerRoundTrips()
    {
        var rc = new RemoteConfig(Url, new RemoteAuth.Bearer("secret-tok"), RemoteLaunchStyle.Npx);
        var encoded = RemotePattern.Encode(rc);
        Assert.Equal(["-y", "mcp-remote", Url, "--header", "Authorization:${AUTH_HEADER}"], Args(encoded)!);
        Assert.Equal(Pairs(("AUTH_HEADER", "Bearer secret-tok")), Env(encoded));
        Assert.Equal(rc, RemotePattern.Decode(encoded));
    }

    [Fact]
    public void HeaderRoundTrips()
    {
        var rc = new RemoteConfig(Url, new RemoteAuth.Header("X-API-Key", "k123"), RemoteLaunchStyle.Npx);
        var encoded = RemotePattern.Encode(rc);
        Assert.Equal(["-y", "mcp-remote", Url, "--header", "X-API-Key:${AUTH_HEADER}"], Args(encoded)!);
        Assert.Equal(Pairs(("AUTH_HEADER", "k123")), Env(encoded));
        Assert.Equal(rc, RemotePattern.Decode(encoded));
    }

    [Fact]
    public void OAuthClientRoundTrips()
    {
        var rc = new RemoteConfig(Url, new RemoteAuth.OAuthClient("cid", "csecret", "read write"), RemoteLaunchStyle.Npx);
        var encoded = RemotePattern.Encode(rc);
        Assert.Equal(
            ["-y", "mcp-remote", Url,
             "--static-oauth-client-info", "{\"client_id\":\"cid\",\"client_secret\":\"csecret\"}",
             "--static-oauth-client-metadata", "{\"scope\":\"read write\"}"],
            Args(encoded)!);
        Assert.Equal(rc, RemotePattern.Decode(encoded));
    }

    // specific decode scenarios

    [Fact]
    public void DecodeBearerFromExplicitConfig()
    {
        var config = NpxConfig(["-y", "mcp-remote", Url, "--header", "Authorization:${AUTH_HEADER}"], Pairs(("AUTH_HEADER", "Bearer abc")));
        var decoded = RemotePattern.Decode(config);
        Assert.NotNull(decoded);
        Assert.Equal(new RemoteAuth.Bearer("abc"), decoded.Auth);
        Assert.Empty(decoded.PassthroughEnv);
        Assert.Equal(config, RemotePattern.Encode(decoded));
    }

    [Fact]
    public void DecodeCustomHeaderFromExplicitConfig()
    {
        var config = NpxConfig(["-y", "mcp-remote", Url, "--header", "X-API-Key:${AUTH_HEADER}"], Pairs(("AUTH_HEADER", "topsecret")));
        var decoded = RemotePattern.Decode(config);
        Assert.NotNull(decoded);
        Assert.Equal(new RemoteAuth.Header("X-API-Key", "topsecret"), decoded.Auth);
        Assert.Equal(config, RemotePattern.Encode(decoded));
    }

    [Fact]
    public void OAuthClientWithoutScopes()
    {
        var rc = new RemoteConfig(Url, new RemoteAuth.OAuthClient("cid", "", ""), RemoteLaunchStyle.Npx);
        var encoded = RemotePattern.Encode(rc);
        Assert.Equal(["-y", "mcp-remote", Url, "--static-oauth-client-info", "{\"client_id\":\"cid\",\"client_secret\":\"\"}"], Args(encoded)!);
        Assert.Equal(rc, RemotePattern.Decode(encoded));
    }

    // preservation

    [Fact]
    public void BearerPlusExtraHeaderPreservesBoth()
    {
        var config = NpxConfig(
            ["-y", "mcp-remote", "https://x.dev/mcp", "--header", "Authorization:${AUTH_HEADER}", "--header", "X-Tenant:acme"],
            Pairs(("AUTH_HEADER", "Bearer secret-tok")));
        var decoded = RemotePattern.Decode(config);
        Assert.NotNull(decoded);
        Assert.Equal(new RemoteAuth.Bearer("secret-tok"), decoded.Auth);
        Assert.Equal(["--header", "X-Tenant:acme"], decoded.ExtraArgs);
        Assert.Empty(decoded.PassthroughEnv);
        var re = RemotePattern.Encode(decoded);
        var a = Args(re) ?? [];
        Assert.Contains("Authorization:${AUTH_HEADER}", a);
        Assert.Contains("X-Tenant:acme", a);
        Assert.Equal("Bearer secret-tok", Env(re)["AUTH_HEADER"]);
    }

    [Fact]
    public void UnrecognizedHeaderEnvVarIsPreservedNotFabricated()
    {
        var config = NpxConfig(["-y", "mcp-remote", "https://x.dev/mcp", "--header", "X-Key:${MY_KEY}"], Pairs(("MY_KEY", "abc")));
        var decoded = RemotePattern.Decode(config);
        Assert.NotNull(decoded);
        Assert.Equal(RemoteAuth.Auto, decoded.Auth);
        Assert.Equal(["--header", "X-Key:${MY_KEY}"], decoded.ExtraArgs);
        Assert.Equal("abc", decoded.PassthroughEnv["MY_KEY"]);
        Assert.Equal(config, RemotePattern.Encode(decoded));
    }

    [Fact]
    public void ExtraArgsPreserved()
    {
        var rc = new RemoteConfig(Url, RemoteAuth.Auto, RemoteLaunchStyle.Npx, ["--transport", "http-only"]);
        var encoded = RemotePattern.Encode(rc);
        Assert.Equal(["-y", "mcp-remote", Url, "--transport", "http-only"], Args(encoded)!);
        Assert.Equal(rc, RemotePattern.Decode(encoded));
    }

    [Fact]
    public void PassthroughEnvSurvivesDecodeThenEncode()
    {
        var config = NpxConfig(["-y", "mcp-remote", Url], Pairs(("UNRELATED", "keep-me")));
        var decoded = RemotePattern.Decode(config);
        Assert.NotNull(decoded);
        Assert.Equal(Pairs(("UNRELATED", "keep-me")), decoded.PassthroughEnv);
        Assert.Equal(RemoteAuth.Auto, decoded.Auth);
        Assert.Equal(config, RemotePattern.Encode(decoded));
    }

    // versioned marker

    [Fact]
    public void VersionedMarkerDecodes()
    {
        var decoded = RemotePattern.Decode(NpxConfig(["-y", "mcp-remote@latest", Url]));
        Assert.Equal(Url, decoded?.Url);
        Assert.Equal(RemoteAuth.Auto, decoded?.Auth);
    }

    // rejection

    [Fact]
    public void DecodeReturnsNullForNonNpxCommand()
    {
        var config = JsonValue.Object(("command", JsonValue.String("node")),
            ("args", JsonValue.Array([JsonValue.String("-y"), JsonValue.String("mcp-remote"), JsonValue.String(Url)])));
        Assert.Null(RemotePattern.Decode(config));
    }

    [Fact]
    public void DecodeReturnsNullForNonMarker()
    {
        Assert.Null(RemotePattern.Decode(NpxConfig(["-y", "some-other-package", Url])));
    }

    [Fact]
    public void DecodeReturnsNullForMissingUrl()
    {
        Assert.Null(RemotePattern.Decode(NpxConfig(["-y", "mcp-remote"])));
    }
}
